File: py/50.pow-x-n.py
#
# @lc app=leetcode.cn id=50 lang=python3
#
# [50] Pow(x, n)
#

# @lc code=start
class Solution:
    # 迭代
    # 56ms    5%    5%    13.6MB
    def myPow1(self, x: float, n: int) -> float:
        res = 1
        if n ==0 :
            return res
        pn = n
        if pn < 0:
            pn = -pn
        loop_res = x
        while pn > 0:
            if pn%2 == 1:
                res = res*loop_res
            loop_res = loop_res * loop_res
            pn = int(pn/2)
        if n < 0:
            res = 1/res
        return res

# 递归法
# 56ms   5%   5%   13.7MB
    def myPow2(self, x: float, n: int) -> float:
        res = 1
        if n ==0 :
            return res
        pn = n
        if pn < 0:
            pn = -pn
        def divide(n):
            if n == 1:
                return x
            divide_res = divide(int(n/2))
            return divide_res*divide_res if n%2==0 else divide_res * divide_res* x
        res = divide(pn)

        return res if n>0 else 1/res

# 递归法，使用位运算代替除法和偶数判断
# 40ms    72%   41%   13,4MB
    def myPow3(self, x: float, n: int) -> float:
        res = 1
        if n ==0 :
            return res
        pn = n
        if pn < 0:
            pn = -pn
        def divide(n):
            if n == 1:
                return x
            divide_res = divide(n>>1)
            return divide_res*divide_res if n&1==0 else divide_res * divide_res* x
        res = divide(pn)

        return res if n>0 else 1/res
    
        # 迭代
    # 40ms    72%    5%    13.7MB
    def myPow(self, x: float, n: int) -> float:
        res = 1
        if n ==0 :
            return res
        pn = n
        if pn < 0:
            x = 1/x
            pn = -pn
        loop_res = x
        while pn > 0:
            if pn&1:
                res = res*loop_res
            loop_res = loop_res * loop_res
            pn = pn >> 1
        return res
# @lc code=end

File: java/50.pow-x-n.java
/*
 * @lc app=leetcode.cn id=50 lang=java
 *
 * [50] Pow(x, n)
 */

// @lc code=start
class Solution {
    /**
     * 暴力法，n 次方就相乘n次
     * time limit exceeded
     */
    public double myPow1(double x, int n) {
        double res =  1;
        if(n==0) return res;
        if(x==1.0) return x;
        int pn = n;
        if(pn<0) pn = -pn;
        
        for(int i=0;i<pn;i++){
            res = x * res;
        }
        if(n<0) res = 1/res;
        return res;
    }

    /**
     * 1ms   97%   61%   36.3MB
     * 递归，分治
     * 分治结果，根据传入的偶数还是奇数，采用res * res 或者res*res*x
     */
    public double myPow2(double x, int n) {
        double res =  1;
        if(n==0) return res;
        long pn = n;
        if(pn<0) pn = -pn;
        
        res = divide(x,pn);
        if(n<0) res = 1/res;
        return res;
    }
    private double divide(double x,long n){
        if(n == 1) return x;
        long leftN = n/2;
        double leftRes = divide(x,leftN);
        return n%2 == 0 ?leftRes* leftRes:leftRes*leftRes*x;
    }

    /**
     * 迭代法
     * 1ms    97%    12%    37.9MS
     */
    public double myPow(double x, int n) {
        double res =  1;
        if(n==0) return res;
        long pn = n;
        if(pn<0) pn = -pn;
         double loopRes = x;
        // res = x;
        
        while(pn > 0){
            if(pn%2==1){
                res = res * loopRes;
            }
            loopRes = loopRes * loopRes;
            pn = pn/2;
        }
        if(n<0) res = 1.0/res;
        return res;
    }
}
// @lc code=end

